[DSpace-CRIS] Creation of a custom URL for Items (Backend)

AdamF42 opened the pull request:
## References
* Related to [RestContract#318](https://github.com/DSpace/RestContract/pull/318)
* Realted to [Dspace-Angular#4814](https://github.com/DSpace/dspace-angular/pull/4814)

## Description
This PR migrates the custom URL feature from DSpace-CRIS to core DSpace and refactors the validation system to adopt the DSpace-CRIS validation architecture, moving validation logic from the REST layer to the API layer for better reusability and maintainability.

## Instructions for Reviewers
This PR brings two major improvements to DSpace: **Custom URL support** and **Validation system modernization**. The custom URL feature allows items to have human-friendly URLs instead of UUID-based paths, while the validation refactor moves from REST-coupled validation to a cleaner API-layer architecture.

### List of changes in this PR:

**Custom URL Feature:**
* **New Service Layer**: Added `CustomUrlService` interface and `CustomUrlServiceImpl` for managing custom URLs
* **Metadata Support**: Added `dspace.customurl` and `dspace.customurl.old` fields to `dspace-types.xml` 
* **Automatic Generation**: Added `CustomUrlConsumer` that auto-generates URLs from item titles during submission
* **Search Integration**: Added `SolrServiceCustomUrlIndexPlugin` for indexing custom URLs in Solr
* **REST API**: Added `CustomUrlStep` for submission workflow and PATCH operations for URL management
* **Item Lookup**: Added `findByCustomUrl()` method to `ItemRestRepository`
* **Configuration**: Added custom URL consumer to `dspace.cfg` event dispatcher

**Validation System Refactor:**
* **Architecture Migration**: Moved validation from `dspace-server-webapp` to `dspace-api` package
* **New Framework**: Created `ValidationService`, `SubmissionStepValidator`, and `GlobalSubmissionValidator` interfaces
* **Migrated Validators**: 
  - `MetadataValidator` (from `MetadataValidation`)
  - `LicenseValidator` (from `LicenseValidator`)
  - `CclicenseValidator` (from `CclicenseValidator`)
  - `UploadValidator` (from `UploadValidation`)
  - `NotifyValidator` (from `NotifyValidation`)
  - `CustomUrlValidator` (new for URL validation)
* **Service Layer**: Added validation service factory and implementation with dependency injection
* **Error Model**: New structured `ValidationError` model for consistent error reporting
* **Spring Configuration**: Added `addon-validation-services.xml` for validation service beans

**Technical Improvements:**
* **Utilities**: Added `TypeBindUtils` for submission type binding functionality

### How to test:

Use this [frontend](https://github.com/DSpace/dspace-angular/pull/4814). 

**Testing Custom URLs:**
1. **Enable Feature**: Ensure `customurl` consumer is in `event.dispatcher.default.consumers` in `dspace.cfg` and the property `dspace.custom-url.consumer.supported-entities` and `dspace.custom-url.consumer.entity-metadata-mapping` are set to a valid entity (es. Publication) and valid metadata. (see the example in dspace.cfg).
2. **Create an Item**: Submit a new item (es. Publication) through the submission workflow
3. **Verify Auto-Generation**: Check that a custom URL is automatically generated from the item title
4. **Test Manual URLs**: Use the Person submission interface to set custom URLs manually
5. **Test Lookup**: Verify items can be found using custom URLs via REST API

**Testing Validation:**
1. **Submission Workflow**: Test that all existing validation still works (required fields, file uploads, licenses)
2. **Custom URL Validation**: Try setting duplicate or invalid custom URLs to verify validation errors
3. **Step Navigation**: Ensure validation doesn't block legitimate workflow progression

**Testing Migration:**
1. **Backward Compatibility**: Verify existing items and submissions continue to work unchanged
2. **Configuration**: Test with both enabled and disabled custom URL consumer

## Checklist
- [x] My PR is **created against the `main` branch** of code (unless it is a backport or is fixing an issue specific to an older branch).
- [ ] My PR is **small in size** (e.g. less than 1,000 lines of code, not including comments & integration tests). Exceptions may be made if previously agreed upon.
- [x] My PR **passes Checkstyle** validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide).
- [x] My PR **includes Javadoc** for _all new (or modified) public methods and classes_. It also includes Javadoc for large or complex private methods.
- [x] My PR **passes all tests and includes new/updated Unit or Integration Tests** based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] My PR **includes details on how to test it**. I've provided clear instructions to reviewers on how to successfully test this fix or feature.
- [x] If my PR includes new libraries/dependencies (in any `pom.xml`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [x] If my PR modifies REST API endpoints, I've opened a separate [REST Contract](https://github.com/DSpace/RestContract/blob/main/README.md) PR related to this change.
- [x] If my PR includes new configurations, I've provided basic technical documentation in the PR itself.
- [ ] If my PR fixes an issue ticket, I've [linked them together](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).

AdamF42 committed:
[DURACOM-412] Added dspace-cris url validator

AdamF42 committed:
[DURACOM-412] Removed dspace validators

AdamF42 committed:
Merge remote-tracking branch 'gh/main' into task/main/DURACOM-412

AdamF42 committed:
[DURACOM-412] Added missing custom url metadata

AdamF42 committed:
[DURACOM-412] Fix on withdrawn items

AdamF42 committed:
[DURACOM-412] Updated javadoc

AdamF42 committed:
Revert "[DURACOM-412] Added missing custom url metadata"

This reverts commit f89c01eacda7d83bfc8fdc71a7d42eab1e491a7d.

AdamF42 committed:
[DURACOM-412] Add dspace.customurl metadata

AdamF42 committed:
[DURACOM-412] Updated new metadata references

AdamF42 committed:
[DURACOM-412] Fix on MetadataValidator

AdamF42 committed:
[DURACOM-412] Fix on missing notify validator

AdamF42 committed:
[DURACOM-412] updated dspace.cfg

kskaiser commented on dspace-api/src/main/java/org/dspace/app/customurl/service/CustomUrlServiceImpl.java:
```diff
@@ -0,0 +1,187 @@
+
+        if (indexableObjects.size() > 1) {
+            LOGGER.error("Found many item with the same custom url {} - Ids: {}", customUrl, getIds(indexableObjects));
+            throw new IllegalStateException("Found many item with the same custom url: " + customUrl);
```
The log message and exception text could be phrased a bit more clearly. For example, “Found many item” should be “Found multiple items”, and using consistent capitalization for “URL” improves readability. It might also help to make the messages more precise for troubleshooting.

AdamF42 committed:
Merge remote-tracking branch 'gh/main' into task/main/DURACOM-412

nwoodward commented on dspace/config/dspace.cfg:
```diff
@@ -1452,7 +1480,7 @@ sitemap.dir = ${dspace.dir}/sitemaps
 # Customize the path of sitemaps in the server webapp
 # Defaults to "sitemaps", which means they are available at ${dspace.server.url}/sitemaps/
-# sitemap.path = sitemaps
+sitemap.path = sitemaps
```
Was this change to `sitemap.path` supposed to be in the PR?

AdamF42 committed:
[DURACOM-412] commented "sitemap.path = sitemaps" on dspace.cfg

AdamF42 committed:
[DURACOM-412] Updated CustomUrlRemoveOperation to extend RemovePatchOperation instead of ReplacePatchOperation

AdamF42 committed:
[DURACOM-412] empty commit to trigger pipeline

AdamF42 committed:
Merge remote-tracking branch 'main' into task/main/DURACOM-412

kskaiser reviewed (changes requested):
I reviewed this together with https://github.com/DSpace/dspace-angular/pull/4814

Good work, but I noticed some problems that need to be addressed.
Among the issues raised by @nwoodward about the possibility to create duplicate customurls manually, the duplicates can also be automatically created via the CustomUrlConsumer.
So on the one hand, when an item with the same "dc.title" field as an existing one (can also be another entity type) is created, the customurl consumer will create the same customurl.
Furthermore, also when creating a new version of an item (but with the same "dc.title" value), again a customurl field with the same value is created.
Unfortunately, then none of the items or item's versions can be accessed anymore.

I also checked the test coverage. This is quite good for the consumer class, but should be improved for other classes. 

Would it be possible to clean up the commit history before merging (squash/interactive rebase)? That would simplify later tracing.

AdamF42 committed:
[DURACOM-412] Enhanced the `CustomUrlConsumer` class to support configurable metadata fields for generating custom URLs based on entity types.

AdamF42 committed:
[DURACOM-412] Implemented a solution to handle duplicate custom URLs by generating progressive version

AdamF42 commented:
> > I reviewed this together with [DSpace/dspace-angular#4814](https://github.com/DSpace/dspace-angular/pull/4814)
> > Good work, but I noticed some problems that need to be addressed. Among the issues raised by @nwoodward about the possibility to create duplicate customurls manually, the duplicates can also be automatically created via the CustomUrlConsumer. So on the one hand, when an item with the same "dc.title" field as an existing one (can also be another entity type) is created, the customurl consumer will create the same customurl. Furthermore, also when creating a new version of an item (but with the same "dc.title" value), again a customurl field with the same value is created. Unfortunately, then none of the items or item's versions can be accessed anymore.
> > I also checked the test coverage. This is quite good for the consumer class, but should be improved for other classes.
> > Would it be possible to clean up the commit history before merging (squash/interactive rebase)? That would simplify later tracing.
> 
> Thanks for the feedback!
> 
> Regarding the test coverage: classes such as `*PatchOperation` and `CustomUrlServiceImpl` are indirectly covered by the `CustomUrlStepIT` integration tests. However, if you think it would be helpful, I can also provide dedicated unit tests for `CustomUrlServiceImpl`.
> 
> As for the commit history, since there are several intermediate merges, I’d prefer to avoid squashing unless it’s strictly necessary.
> 
> About the duplicate issue: We propose adding a logic to check the custom URL existence when a new one is generated, and if the same value already exists, automatically create a progressive version (e.g., `customurl-1`, `customurl-2`, etc.). In any case, administrators will still be able to manually modify the value, and future PRs introducing the edit mode (with validation) will ensure proper consistency.
> 
> Would this approach work for you?


@kskaiser I've implemented duplicate handling with progressive versioning (`john-doe` → `john-doe-1` → `john-doe-2`).

**Versioned Items Issue**
Current `versioning-service.xml` excludes `dspace.customurl`, so:
- New versions don't inherit the URL → Consumer generates a new one
- **If Consumer is disabled/not configured** → **No URL on new version**

Based on our experience, item versioning is rarely used, so this edge case wasn't noticed during our original implementation.

**Question**: Should versioned items get incremental URLs or get the same URL?

**Proposal**:
1. **Delete** `dspace.customurl` from old version in `DefaultItemVersionProvider#updateItemState`
2. **Skip generation** on old versions using `VersionHistoryServiceImpl#isLastVersion()` in `CustomUrlConsumer`
3. New version **inherits/reuses** the URL

Does this solution work for you?

AdamF42 committed:
[DURACOM-412] Assign custom URL only to latest item version and clear from previous and improved tests

AdamF42 commented:
@kskaiser Thanks for the feedback!  
I've implemented the agreed approach so that only the **latest version** of an item retains the primary `dspace.customurl`.

### Summary of Updates
1. **Custom URL Rotation in `AbstractVersionProvider`**
   - Introduced a new `rotateCustomUrl` method to manage URL transitions during versioning.
   - The **old version** now moves its `dspace.customurl` to `dspace.customurl.old` to preserve history and clears the active field.
   - The **new version** inherits the original URL through the normal metadata copy process, ensuring that the primary URL always resolves to the most recent version.

2. **Authorization Handling**
   - Since archived items may restrict write operations, the rotation logic safely uses `context.turnOffAuthorisationSystem()`, but only after confirming that the user has permission to create a new version via `itemService.canCreateNewVersion()`.

3. **Consumer Behavior**
   - The URL is cleared from the old version **before** completing the transaction, allowing the `CustomUrlConsumer` to treat it as available and preventing unnecessary suffixed URLs.

4. **Additional Tests**
   - As requested, I’ve added more specific tests in `CustomUrlServiceImplIT` to cover the key methods.

benbosman commented:
Hi @AdamF42 ,
I was just looking at this PR, and I was wondering why it's not using the [existing identifiers endpoint](https://github.com/DSpace/RestContract/blob/main/identifiers.md)? Wouldn't that already support a lot of the requirements, eliminating the need for new REST endpoints

benbosman reviewed (commented):
Thanks for the PR
I've included some feedback inline

AdamF42 committed:
[DURACOM-412] Optimize CustomUrlConsumer performance and fix URL validation pattern

- Cache entity-to-metadata mapping in CustomUrlConsumer to avoid repeated configuration loading
- Fix CustomUrlValidator regex pattern to properly reject URLs with forward slashes
- Add comprehensive tests for invalid URL validation and private item access control

AdamF42 committed:
[DURACOM-412] Fixed CustomUrlConsumerIT

tdonohue commented on dspace-server-webapp/src/test/java/org/dspace/app/rest/ItemRestRepositoryIT.java:
```diff
@@ -4902,4 +4902,195 @@ public void findSubmitterByAnonymousTest() throws Exception {
+                        .andExpect(status().isOk())
+                        .andExpect(jsonPath("$.uuid", is(item.getID().toString())));
+    }
+
```
I think we need to add a test that `findByCustomURL` still works after an Item is Versioned.  After looking at the code, I believe there may be a flaw, because a versioned Item appears to be updated like this:
* New version has custom URL in `dspace.customurl`
* Prior version has custom URL in `dspace.customurl.old`
* However, the flaw is that `findByCustomURL` searches across BOTH `dspace.customurl` and `dspace.customurl.old` (as both of these fields are indexed into Solr as `customurl`).

So, I believe a versioned item *cannot* use `dspace.customurl.old` for old versions, as that duplicates the custom URL, resulting in `findByCustomURL` throwing an error.

Alternatively, we could reconsider whether to keep old custom URLs in the `dspace.customurl.old` altogether.  I realize they are there to help redirect old custom urls, but these are NOT persistent identifiers...so the redirects may be difficult to "persist" across versions or in other scenarios.

tdonohue replied:
I see this test was added to `CustomUrlConsumerIT.testItemVersioningCausesCustomUrlSearch()`

AdamF42 replied:
Yes, I added `testItemVersioningCausesCustomUrlSearch` to `CustomUrlConsumerIT` to tests the full versioning flow. 

I've fixed the versioning issue with the following changes:

**1. Solr Search Filtering (`latestVersion:true`)**
Added `latestVersion:true` filter to all Solr queries in `CustomUrlServiceImpl` because the search was finding both old and new versions when they share the same custom URL value. 

When versioning occurs:
- Original item has URL moved to `dspace.customurl.old` (still indexed as `customurl` field)
- New version gets URL in `dspace.customurl`
- Both get indexed to Solr's `customurl` field
- Without the filter, `findItemByCustomUrl` throws "multiple items found" error

The filter ensures only the latest version is returned, preventing conflicts.

**2. Validation Fix (`CustomUrlValidator`)**
Updated `CustomUrlValidator` to use `findArchivedByMetadataFieldExcludingOldVersions()`.

The validator was incorrectly blocking new versions from reusing custom URLs because it found old versions as "duplicates". Now it only validates against latest versions, allowing version chains to reuse URLs.

tdonohue reviewed (commented):
@AdamF42 and @FrancescoMolinaro :  I've only partially reviewed this PR.  But, I think I've found the source of the Item Versioning bug described in [my review on the frontend](https://github.com/DSpace/dspace-angular/pull/4814#pullrequestreview-3729593466).  I've noted it inline below.

Also a few other minor cleanup things that I saw.  Overall, the code looks good, but I'm only roughly 1/2 way through the code review.  I hope to finish it on Monday.

AdamF42 committed:
[DURACOM-412] Address PR review feedback for Custom URL feature (#11549)
- Fix dspace-types.xml metadata field descriptions:
  * Correct scope_note to indicate Custom URLs work for DSpace entities only
  * Fix capitalization: "Dspace" → "DSpace"
- Disable custom-url step by default in item-submission.xml:
  * Add explanatory comment about enabling the feature
  * Comment out <step id="custom-url"/> in Person submission process
- Change metadata field separator from semicolon to comma:
  * Update dspace.cfg documentation and examples
  * Refactor CustomUrlConsumerConfig to use getArrayProperty()
  * Update all CustomUrlConsumerIT test cases to use comma separator

AdamF42 committed:
[DURACOM-412] Updated versioning support to Custom URL feature
Implement comprehensive versioning support ensuring custom URLs work correctly with versioned items. Old versions no longer interfere with custom URL resolution, validation, or uniqueness.
Changes:
- Add findByMetadataFieldExcludingOldVersions() to ItemDAO/ItemService with query to filter non-latest versions
- Update CustomUrlValidator to use new method
- Add latestVersion:true filter to all Solr searches in CustomUrlServiceImpl  (findItemByCustomUrl, customUrlExists, findCustomUrlsWithPattern)
- Update ItemRestRepository.findByCustomUrl() to return HTTP 404 instead of empty body when item not found

AdamF42 commented on dspace-api/src/main/java/org/dspace/app/customurl/consumer/CustomUrlConsumerConfig.java:
```diff
@@ -0,0 +1,152 @@
+     * @return list of trimmed, non-empty metadata field names
+     */
+    private List<String> parseMetadataFieldList(String metadataFieldsConfig) {
+        return Arrays.stream(metadataFieldsConfig.split(METADATA_FIELD_SEPARATOR))
```
Thank you for the suggestion. I've updated the configuration to use comma-separated values instead of semicolons. The `CustomUrlConsumerConfig` now uses `configurationService.getArrayProperty()`.

tdonohue reviewed (approved):
👍 Thanks @AdamF42 !  I've finished my code review and the rest of the code looks fine.  I've also retested alongside https://github.com/DSpace/dspace-angular/pull/4814 .  With your recent updates, I've verified that the [bugs I previously reported](https://github.com/DSpace/dspace-angular/pull/4814#pullrequestreview-3729593466) are now all fixed.  This includes the incompatibility with Item Versioning and the update to return a 404 if a custom URL is not found.

I've also looked back on the detailed feedback from @kskaiser , @nwoodward , and @benbosman .  They should correct me if I'm wrong, but I **believe** all their feedback has been addressed, **except for one severe bug which will be delayed for a future PR**:
* As both @kskaiser and @nwoodward reported, it's possible to manually create duplicate "custom URLs" using the Edit Metadata page on an Item.  When that is done, the "custom URL" fails to resolve.  --> UPDATE, moved to https://github.com/DSpace/dspace-angular/issues/5078

Overall, though, great work. This looks good to me.

AdamF42 commented:
> @AdamF42 , were the following open bug tickets from the CRIS project (backend) taken into account during the implementation? Will the bug fixes be backported to CRIS?
> 
>     * [Custom URL with slash makes the item unaccessible 4Science/DSpace#403](https://github.com/4Science/DSpace/issues/403)
> 
>     * [For items with a custom URL it's impossible to upload files to a new bundle 4Science/DSpace#372](https://github.com/4Science/DSpace/issues/372)
> 
> 
> There is also an open PR: [4Science#408](https://github.com/4Science/DSpace/pull/408)

@saschaszott Thank you for bringing up these issues!
1. **Issue https://github.com/4Science/DSpace/issues/403** (Custom URLs with slashes): This has been addressed. The `CustomUrlValidator` now returns an error when a custom URL includes slashes, preventing the routing conflicts described in that issue.
2. **Issue https://github.com/4Science/DSpace/issues/372** (Administrative Edit mode validation):  This issue is currently related to Angular cache behavior, not backend validation. The problem affects items with and without custom URLs - doing a browser refresh with cache disabled shows the created bundle correctly. Since this is an Angular-side caching issue that affects the broader edit functionality, it's not related to the current pull request.

4. **PR https://github.com/4Science/DSpace/pull/408**: This PR is no longer valid. We have comprehensively handled custom URL behavior in versioning scenarios as explained in previous comments. The current implementation ensures that only the latest version of an item retains the primary custom URL, while older versions move their URLs to the historical field.

AdamF42 committed:
[DURACOM-412] Improved tests to check retrieval of non discoverable items

tdonohue reviewed (approved):
👍 Thanks @AdamF42 !  I retested today alongside the frontend and found no issues. This still looks good to me.

@nwoodward and @kskaiser and @benbosman  : You all have submitted change requests on this PR.  While I *believe* the requested changes have now been made, I'd appreciate it if you can verify that on your end soon.  I believe this PR is ready to merge, but it needs another +1.  (I'd like to merge this quickly, if possible, to make the larger "Authority Framework" PR #11718 easier to review.)

tdonohue commented:
As this is now at +3, merging this PR.  Thanks again @AdamF42 !

If anyone finds that any feedback on this PR has not been addressed (or find undiscovered bugs later), then please create a separate ticket and we'll work to get it resolved in a follow-up PR.

GraziaQuercia commented:
documentation ready https://wiki.lyrasis.org/display/DSDOC10x/Custom+URL+for+Entities
